pyrate/core/dem_error.py

```diff
@@ -0,0 +1,307 @@
+
+    return lon, lat
+
+
+def get_radar_coords(ifg, params):
+    """
+    Function to get radar coordinates for each pixel in the multi-looked interferogram dataset.
+    Radar coordinates are identical for each interferogram in the stack.
+    """
+    # lookup table file:
+    lookup_table = params[cf.LT_FILE]
+    # PyRate IFG multi-looking factors
+    ifglksx = params[cf.IFG_LKSX]
+    ifglksy = params[cf.IFG_LKSY]
+    print(lookup_table)
+    # transform float lookup table file to np array
+    lt_az, lt_rg = read_lookup_table(ifg, lookup_table, ifglksx, ifglksy)
+    # replace 0.0 with NaN
+    lt_az[lt_az==0.0] = np.nan
+    lt_rg[lt_rg == 0.0] = np.nan
+
+    return lt_az, lt_rg
+
+
+def calc_local_geometry(ifg, ifg_path, az, rg, lon, lat, params):
+    """
+    Function to calculate local look angle, incidence angle and geodetic azimuth for each pixel.
+    """
+    # read relevant metadata from first IFG
+    a = float(ifg.meta_data[ifc.PYRATE_SEMI_MAJOR_AXIS_METRES])
+    b = float(ifg.meta_data[ifc.PYRATE_SEMI_MINOR_AXIS_METRES])
+    se = float(ifg.meta_data[ifc.PYRATE_SAR_EARTH_METRES])
+    near_range = float(ifg.meta_data[ifc.PYRATE_NEAR_RANGE_METRES])
+    rps = float(ifg.meta_data[ifc.PYRATE_RANGE_PIX_METRES])
+    heading = float(ifg.meta_data[ifc.PYRATE_HEADING_DEGREES])
+
+    # convert angles to radians
+    lon = lon / 180 * np.pi
+    lat = lat / 180 * np.pi
+    heading = heading / 180 * np.pi
+
+    # Earth radius at given latitude
+    re = np.sqrt(np.divide(np.square(a**2 * np.cos(lat)) + np.square(b**2 * np.sin(lat)), \
+                           np.square(a * np.cos(lat)) + np.square(b * np.sin(lat))))
+
+    # range measurement at pixel ij -> this requires the mli.par as input since rps would be different in the slc.par
+    range = near_range + rps*rg
+
+    # look angle at pixel ij -> law of cosines in "satellite - Earth centre - ground pixel" triangle
+    # see e.g. Section 2 in https://www.cs.uaf.edu/~olawlor/ref/asf/sar_equations_2006_08_17.pdf
+    look_angle = np.arccos(np.divide(se**2 + np.square(range) - np.square(re), 2 * se * range))
+
+    # incidence angle at pixel ij -> law of cosines in "satellite - Earth centre - ground pixel" triangle
+    # see e.g. Section 2 in https://www.cs.uaf.edu/~olawlor/ref/asf/sar_equations_2006_08_17.pdf
+    incidence_angle = np.pi - np.arccos(np.divide(np.square(range) + np.square(re) - se**2, 2 * np.multiply(range, re)))
+    # for validation with GAMMA output only
+    incidence_angle_gamma = np.pi / 2 - incidence_angle
+    # maximum differences to the GAMMA-derived local incidence angles for Sentinel-1 test data are within +/-0.1 deg
+
+    # local azimuth angle at pixel ij using constant satellite heading angle and spherical approximations
+    epsilon = np.pi - look_angle - (np.pi - incidence_angle) # angle at the Earth's center between se and re
+    # azimuth of satellite look vector(for right-looking SAR)
+    sat_azi = heading + np.pi / 2
```

Comment:
as discussed, read this from the Gamma header file?

Reply:
Implemented reading of parameter `azimuth_angle` from GAMMA mli.par file and using its value (i.e. +90 deg or pi/2 for right-looking SAR sensors) instead of restricting the equation to right-looking SAR sensors.